### PR TITLE
Fixed bubble rendering issue, added ConfigurableZone

### PR DESCRIPTION
**Pull request by AisenGinn**
Fixed Envision bubble rendering issue, added ConfigurableZone in to bubble.

### CODE REVIEW

File: smarts/sstudio/types.py [hunk 1]
```diff
@@ -813,6 +813,36 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
             poly = shapely_rotate(poly, self.rotation, use_radians=True)
         return shapely_translate(poly, xoff=x, yoff=y)
 
+@dataclass(frozen=True)
+class ConfigurableZone(Zone):
+    """A descripter that defines a specific configurableZone defined by user"""
+
+    ext_coordinates: List[Tuple[float, float]]
+    """external coordinates of the polygon"""
```
Comment (Gamenot) [inline smarts/sstudio/types.py:821]: Please describe how many points are valid for what kind of shape is desired. 

- <2 points causes an error
- 2 points generates a box
- \>2 points generates a shape

File: smarts/sstudio/types.py [hunk 2]
```diff
@@ -813,6 +813,36 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
             poly = shapely_rotate(poly, self.rotation, use_radians=True)
         return shapely_translate(poly, xoff=x, yoff=y)
 
+@dataclass(frozen=True)
+class ConfigurableZone(Zone):
+    """A descripter that defines a specific configurableZone defined by user"""
+
+    ext_coordinates: List[Tuple[float, float]]
```
Comment (Gamenot) [inline smarts/sstudio/types.py:820]: Please add a `__post_init__` that can check that assert that there are a minimum of 2 points provided.

File: smarts/sstudio/types.py [hunk 3]
```diff
@@ -813,6 +813,36 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
...
+        if (
+            len(self.ext_coordinates) == 2
+        ):  # if user only specified two points, create a box
+            x_min = min(self.ext_coordinates[0][0], self.ext_coordinates[1][0])
+            x_max = max(self.ext_coordinates[0][0], self.ext_coordinates[1][0])
+            y_min = min(self.ext_coordinates[1][0], self.ext_coordinates[1][1])
+            y_max = max(self.ext_coordinates[1][0], self.ext_coordinates[1][1])
+            poly = Polygon(
+                [(x_min, y_min), (x_max, y_min), (x_max, y_max), (x_min, y_max)],
+            )
```
Comment (Gamenot) [inline smarts/sstudio/types.py:837]: Will this cause an error if you provide a situation where it describes an axis-parallel line: `p1.x == p2.x` or `p1.y == p2.y`?

The box would have a width or height of 0 then.

**Review by Gamenot — CHANGES_REQUESTED** [review]
Please add a CHANGLOG record.

File: smarts/sstudio/types.py (hunk 3 above)
Comment (AisenGinn) [reply smarts/sstudio/types.py:837]: Maybe I could add the check for this in __post__init__ together with the point numbers check

File: smarts/sstudio/types.py [hunk 4]
```diff
@@ -814,6 +815,51 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
         return shapely_translate(poly, xoff=x, yoff=y)
 
 
+@dataclass(frozen=True)
+class ConfigurableZone(Zone):
+    """A descripter that defines a specific configurableZone defined by user"""
```
Comment (saulfield) [inline smarts/sstudio/types.py:820]:
```suggestion
    """A descriptor for a zone with user-defined geometry."""
```

File: smarts/sstudio/types.py [hunk 5]
```diff
@@ -814,6 +815,51 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
...
+
+    ext_coordinates: List[Tuple[float, float]]
+    """external coordinates of the polygon
+    < 2 points provided: error
+    = 2 points provided: generates a box using these two points as diagonal
+    > 2 points provided: generates a polygons according to the coordinates"""
```
Comment (saulfield) [inline smarts/sstudio/types.py:826]:
```suggestion
    > 2 points provided: generates a polygon according to the coordinates"""
```

File: smarts/sstudio/types.py [hunk 6]
```diff
@@ -814,6 +815,51 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
...
+            raise ValueError(
+                "Two points or more are needed to create a polygon. (less than two points are provided)"
+            )
+
+        x_set = set(point[0] for point in self.ext_coordinates)
+        y_set = set(point[1] for point in self.ext_coordinates)
+        if len(x_set) == 1 or len(y_set) == 1:
+            raise ValueError(
+                "Parallel line cannot form a polygon. (points provided form a parallel line)"
+            )
```
Comment (saulfield) [inline smarts/sstudio/types.py:841]: This message is a bit confusing. I think this could be incorporated into the first check above, just add a check for `len(self.ext_coordinates) < 2`.

File: smarts/sstudio/types.py [hunk 7]
```diff
@@ -814,6 +815,51 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
...
+            y_min = min(self.ext_coordinates[0][1], self.ext_coordinates[1][1])
+            y_max = max(self.ext_coordinates[0][1], self.ext_coordinates[1][1])
+            poly = box(x_min, y_min, x_max, y_max)
+
+        else:  # else create a polygon according to the coordinates
+            poly = Polygon(self.ext_coordinates)
+
+        if self.rotation is not None:
+            poly = shapely_rotate(poly, self.rotation, use_radians=True)
+        return
```
Comment (saulfield) [inline smarts/sstudio/types.py:860]:
```suggestion
        return poly
```

File: smarts/sstudio/types.py [hunk 8]
```diff
@@ -890,6 +936,12 @@ def __post_init__(self):
                 "Only boids can have keep_alive enabled (for persistent boids)"
             )
 
+        poly = None
+        if not isinstance(self.zone, MapZone):
+            poly = self.zone.to_geometry(road_map=None)
+        if poly is not None and (poly.is_valid == False):
+            raise ValueError("The Zone Polygon is not a valid closed loop")
```
Comment (saulfield) [inline smarts/sstudio/types.py:943]: What does it mean if it is `None`? Is that check needed? In any case, this could be simplified:

```python
if not isinstance(self.zone, MapZone):
    poly = self.zone.to_geometry(road_map=None)
    if poly is not None and not poly.is_valid:
        raise ValueError("The Zone Polygon is not a valid closed loop")
```

File: envision/web/src/helpers/state_unpacker.js [hunk 9]
```diff
@@ -33,9 +33,9 @@ const WorldState = Object.freeze({
   SCENARIO_ID: 1,
   SCENARIO_NAME: 2,
   TRAFFIC: 3,
-  BUBBLES: 4,
-  SCORES: 5,
-  EGO_AGENT_IDS: 6,
+  BUBBLES: 5,
+  SCORES: 6,
+  EGO_AGENT_IDS: 7,
```
Comment (saulfield) [inline envision/web/src/helpers/state_unpacker.js:38]: Why was this needed?

**Review by saulfield — CHANGES_REQUESTED** [review]
Looks good, just a few comments.

File: envision/web/src/helpers/state_unpacker.js (hunk 9 above)
Comment (AisenGinn) [reply envision/web/src/helpers/state_unpacker.js:38]: This fixed the bubble rendering issue. There is an empty array in front of the bubbles in 4th position. This causes the bubbles, scores and ego_agent_ids arrays capture the wrong information. (The bubbles capture the empty array, scores are actually bubbles information, and ego_agent_ids are actually scores.)

Comment (AisenGinn) [reply envision/web/src/helpers/state_unpacker.js:38]: I am not sure what the empty array is used for. For now I did not see any usage of that array, but I am not sure whether it contains information in other scenarios, so I just move one position forwards for the bubbles, scores and ego_agent_ids.

File: smarts/sstudio/types.py (hunk 6 above)
Comment (AisenGinn) [reply smarts/sstudio/types.py:841]: This is to check whether user provides multiple coordinates with identical x_axis or y_axis. This will form a parallel line which is not a polygon.

File: smarts/sstudio/types.py (hunk 8 above)
Comment (AisenGinn) [reply smarts/sstudio/types.py:943]: I think you are right, the check for poly is None is not needed, will use the simplified version, thanks!

File: smarts/sstudio/types.py (hunk 6 above)
Comment (saulfield) [reply smarts/sstudio/types.py:841]: Ah yes, I missed the `set`, sorry. I still think we need an explicit check for len < 2 above though.

File: envision/web/src/helpers/state_unpacker.js (hunk 9 above)
Comment (saulfield) [reply envision/web/src/helpers/state_unpacker.js:38]: I just looked into this, and it's because traffic signals were added at index 4 without updating this enum (see `_format_state()` in `envision/data_formatter.py`). I think we should add a line like `SIGNALS: 4,` here so this is clear in the future.

File: smarts/sstudio/types.py [hunk 10]
```diff
@@ -890,6 +936,11 @@ def __post_init__(self):
                 "Only boids can have keep_alive enabled (for persistent boids)"
             )
 
+        if not isinstance(self.zone, MapZone):
+            poly = self.zone.to_geometry(road_map=None)
+            if poly.is_valid:
```
Comment (saulfield) [inline smarts/sstudio/types.py:941]:
```suggestion
            if not poly.is_valid:
```

File: envision/web/src/helpers/state_unpacker.js (hunk 9 above)
Comment (AisenGinn) [reply envision/web/src/helpers/state_unpacker.js:38]: I will add `SIGNALS: 4` in this branch, thanks!

File: smarts/sstudio/types.py (hunk 6 above)
Comment (AisenGinn) [reply smarts/sstudio/types.py:841]: There is a problem of using `len < 2` to check numbers of coordinates. If the user only provides one coordinate, the ext_coordinate will become something like this: ((90,90)). The `len(ext_coordinate)` should be 2 but python somehow unpacks it to be `ext_coordinate = (90,90)` and its `len = 2` again, this may not be a valid check.

Comment (saulfield) [reply smarts/sstudio/types.py:841]: Are you sure? As long as it's always a list it should work. From the REPL:

```
>>> len([(90,90)])
1
```

Comment (AisenGinn) [reply smarts/sstudio/types.py:841]: oh I tried it with this:
```
test = ((90,90))
print(len(test))
```
and it outputs 2

Comment (saulfield) [reply smarts/sstudio/types.py:841]: That would be expected, since you are assigning a tuple of length 2. The outside parentheses are ignored.

```
>>> test = ((90,90))
>>> test
(90, 90)
```

I think we can assume that `ext_coordinates` is always a list.

Comment (AisenGinn) [reply smarts/sstudio/types.py:841]: Normally the user should input a list, but () also works for multiple coordinates, and `len(((90,90),(110,110))) = 2`. Maybe I can convert it to list if index 0 is not a tuple and do the len check?

Comment (saulfield) [reply smarts/sstudio/types.py:841]: This should actually work for tuples as well with no change. But the code would be:

```
>>> test = ((90,90),)
>>> len(test)
1
```

Comment (AisenGinn) [reply smarts/sstudio/types.py:841]: I see, I will add the len check as well, thanks for your patience

**Review by saulfield — APPROVED** [review]

File: smarts/sstudio/types.py [hunk 11]
```diff
@@ -852,6 +853,55 @@ def to_geometry(self, road_map: Optional[RoadMap] = None) -> Polygon:
...
+    = 2 points provided: generates a box using these two points as diagonal
+    > 2 points provided: generates a polygon according to the coordinates"""
+    rotation: Optional[float] = None
+    """The heading direction of the bubble(radians, clock-wise rotation)"""
+
+    def __post_init__(self):
+        if (
+            not self.ext_coordinates
+            or not isinstance(self.ext_coordinates[0], tuple)
+            or len(self.ext_coordinates) < 2
```
Comment (Gamenot) [inline smarts/sstudio/types.py:872]: We should swap these conditions to short circuit and prevent an index error.
```suggestion
            or len(self.ext_coordinates) < 2
            or not isinstance(self.ext_coordinates[0], tuple)
```

File: smarts/sstudio/types.py [hunk 12]
```diff
@@ -928,6 +978,11 @@ def __post_init__(self):
                 "Only boids can have keep_alive enabled (for persistent boids)"
             )
 
+        if not isinstance(self.zone, MapZone):
+            poly = self.zone.to_geometry(road_map=None)
+            if not poly.is_valid:
+                raise ValueError("The Zone Polygon is not a valid closed loop")
```
Comment (Gamenot) [inline smarts/sstudio/types.py:984]: Is there any context we have that we can used to determine which bubble is having an issue?

**Review by Gamenot — CHANGES_REQUESTED** [review]

**Review by Gamenot — APPROVED** [review]

**Comment by Gamenot** [comment]
Good work!
